### PR TITLE
30837 - Bad error message when default 'import'-ing an 'export =' without 'esModuleInterop'

### LINKED ISSUE #30837: Bad error message when default `import`-ing an 'export =' without `esModuleInterop`
Almost related to #13601 and #24542, but not quite.

Run this targeting CommonJS without `esModuleInterop`:

```ts
// ./bar.d.ts
export = bar;
function bar(): void;
```

```ts
// ./foo.ts

import d from "./bar"
```

**Expected**:

```
Module '"c:/Users/Daniel/foo/world"' can only be default-imported using the 'esModuleInterop' flag.
```

**OR**

```
Module '"c:/Users/Daniel/foo/world"' has no default export.
```

with **related span** on the `export =`

```
This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

**Actual**:

```
Module '"c:/Users/Daniel/foo/world"' has no default export.
```

no related span 😢

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #30837

### CODE REVIEW

File: tests/baselines/reference/es6ImportDefaultBindingFollowedWithNamedImportDts.errors.txt [hunk 1]
```diff
@@ -17,26 +17,32 @@ tests/cases/compiler/client.ts(11,8): error TS1192: Module '"tests/cases/compile
 ==== tests/cases/compiler/client.ts (6 errors) ====
     import defaultBinding1, { } from "./server";
            ~~~~~~~~~~~~~~~
-!!! error TS1192: Module '"tests/cases/compiler/server"' has no default export.
+!!! error TS1259: Module '"tests/cases/compiler/server"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/client.ts:1:8: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/es6ImportDefaultBindingFollowedWithNamedImportDts.errors.txt]: None of these was declared with an `export =`.

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -603,10 +603,6 @@
         "category": "Error",
         "code": 1191
     },
-    "Module '{0}' has no default export.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: The solution isn't to delete this message, it's to specialize it when the original module is declared using `export =`

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -2156,6 +2160,10 @@
         "category": "Error",
         "code": 2593
     },
+    "This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.": {
```
Comment (ExE-Boss) [inline src/compiler/diagnosticMessages.json]:
```suggestion
    "This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.": {
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -1953,7 +1953,19 @@ namespace ts {
...
                 if (!exportDefaultSymbol && !hasSyntheticDefault) {
-                    error(node.name, Diagnostics.Module_0_has_no_default_export, symbolToString(moduleSymbol));
+                    if (hasExportAssignmentSymbol(moduleSymbol)) {
+                        const exportEqualsSymbol = moduleSymbol.exports!.get(InternalSymbolName.ExportEquals);
+                        const declaration = exportEqualsSymbol!.valueDeclaration;
+                        const err = error(node.name, Diagnostics.Module_0_can_only_be_default_imported_using_the_esModuleInterop_flag, symbolToString(moduleSymbol));
+
+                        addRelatedInfo(err, createDiagnosticForNode(
+                            declaration,
+                            Diagnostics.This_module_is_declared_with_using_export_and_can_only_be_used_with_a_default_import_when_using_the_esModuleInterop_flag)
```
Comment (ExE-Boss) [inline src/compiler/checker.ts]:
```suggestion
                            Diagnostics.This_module_is_declared_using_export_and_can_only_be_used_with_a_default_import_when_using_the_esModuleInterop_flag)
```

File: tests/baselines/reference/allowSyntheticDefaultImports6.errors.txt [hunk 5]
```diff
@@ -10,6 +10,7 @@ tests/cases/compiler/a.ts(1,8): error TS1192: Module '"tests/cases/compiler/b"'
 ==== tests/cases/compiler/a.ts (1 errors) ====
     import Foo from "./b";
            ~~~
-!!! error TS1192: Module '"tests/cases/compiler/b"' has no default export.
+!!! error TS1259: Module '"tests/cases/compiler/b"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/b.d.ts:4:1: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/allowSyntheticDefaultImports6.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/b.d.ts:4:1: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 6]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
     // default import
     import x1 from "interface";
            ~~
-!!! error TS1192: Module '"interface"' has no default export.
+!!! error TS1259: Module '"interface"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:6:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:6:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 7]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x1 from "interface";
            ~~
-!!! error TS1192: Module '"interface"' has no default export.
+!!! error TS1259: Module '"interface"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:6:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
     import x2 from "variable";
            ~~
-!!! error TS1192: Module '"variable"' has no default export.
+!!! error TS1259: Module '"variable"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:14:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:14:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 8]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x3 from "interface-variable";
            ~~
-!!! error TS1192: Module '"interface-variable"' has no default export.
+!!! error TS1259: Module '"interface-variable"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:26:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:26:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 9]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x4 from "module";
            ~~
-!!! error TS1192: Module '"module"' has no default export.
+!!! error TS1259: Module '"module"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:34:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:34:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 10]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x5 from "interface-module";
            ~~
-!!! error TS1192: Module '"interface-module"' has no default export.
+!!! error TS1259: Module '"interface-module"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:46:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:46:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 11]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x6 from "variable-module";
            ~~
-!!! error TS1192: Module '"variable-module"' has no default export.
+!!! error TS1259: Module '"variable-module"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:60:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:60:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 12]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x7 from "function";
            ~~
-!!! error TS1192: Module '"function"' has no default export.
+!!! error TS1259: Module '"function"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:65:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:65:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 13]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x8 from "function-module";
            ~~
-!!! error TS1192: Module '"function-module"' has no default export.
+!!! error TS1259: Module '"function-module"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:74:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:74:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 14]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x9 from "class";
            ~~
-!!! error TS1192: Module '"class"' has no default export.
+!!! error TS1259: Module '"class"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:82:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:82:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 15]
```diff
@@ -76,34 +76,44 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
     import x0 from "class-module";
            ~~
-!!! error TS1192: Module '"class-module"' has no default export.
+!!! error TS1259: Module '"class-module"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/modules.d.ts:94:5: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/modules.d.ts:94:5: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

File: tests/baselines/reference/es6ImportDefaultBindingInEs5.errors.txt [hunk 16]
```diff
@@ -8,4 +8,5 @@ tests/cases/compiler/es6ImportDefaultBindingInEs5_1.ts(1,8): error TS1192: Modul
 ==== tests/cases/compiler/es6ImportDefaultBindingInEs5_1.ts (1 errors) ====
     import defaultBinding from "./es6ImportDefaultBindingInEs5_0";
            ~~~~~~~~~~~~~~
-!!! error TS1192: Module '"tests/cases/compiler/es6ImportDefaultBindingInEs5_0"' has no default export.
\ No newline at end of file
+!!! error TS1259: Module '"tests/cases/compiler/es6ImportDefaultBindingInEs5_0"' can only be default-imported using the 'esModuleInterop' flag
+!!! related TS2594 tests/cases/compiler/es6ImportDefaultBindingInEs5_0.ts:2:1: This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```
Comment (ExE-Boss) [inline tests/baselines/reference/es6ImportDefaultBindingInEs5.errors.txt]:
```suggestion
!!! related TS2594 tests/cases/compiler/es6ImportDefaultBindingInEs5_0.ts:2:1: This module is declared using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

**Comment by DanielRosenwasser** [comment]
@ExE-Boss is right in that `esModuleInterop` may not be the right flag if your module target is an ECMAScript-style module target.

I think the correct thing to do is to make the message conditional on the module target. `esModuleInterop` for `commonjs` and the like, `allowSyntheticDefaultImports` for `esnext` and `es2015`.

**Comment by DanielRosenwasser** [comment]
@weswigham maybe you should take a look at this too.

**Comment by not-my-profile** [comment]
I want to note that I just found this conditional message to be quite confusing.

> This module is declared with 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.

The "only" here makes the error message look out of date if you're aware of allowSyntheticDefaultImports.

> esModuleInterop may not be the right flag if your module target is an ECMAScript-style module target. I think the correct thing to do is to make the message conditional on the module target. esModuleInterop for commonjs and the like, allowSyntheticDefaultImports for esnext and es2015.

I think it would be nice if this was clearly explained in the [esModuleInterop](https://www.typescriptlang.org/tsconfig#esModuleInterop) documentation because magically changing the option referenced in the error message really doesn't clarify that much.

**Pull request merged**